src/test/scala/firrtlTests/annotationTests/MorphismSpec.scala

```diff
@@ -0,0 +1,570 @@
+// See LICENSE for license details.
+
+package firrtlTests.annotationTests
+
+import firrtl._
+import firrtl.annotations.{Annotation, CircuitTarget, CompleteTarget, DeletedAnnotation}
+import firrtl.annotations.transforms.{DupedResult, ResolvePaths}
+import firrtl.transforms.DedupedResult
+import org.scalatest.{FlatSpec, Matchers}
+
+class MorphismSpec extends FlatSpec with Matchers {
+
+  object AnAnnotation {
+    def apply(target: CompleteTarget) = new AnAnnotation(Some(target))
+  }
+
+  case class AnAnnotation(
+                           target: Option[CompleteTarget],
+                           from: Option[AnAnnotation] = None,
+                           cause: Option[String] = None
+                         ) extends Annotation {
+    override def update(renames: RenameMap): Seq[AnAnnotation] = {
+      if (target.isDefined) {
+        renames.get(target.get) match {
+          case None => Seq(this)
+          case Some(Seq()) => Seq(AnAnnotation(None, Some(this), renames.associatedTransform.map(_.name)))
+          case Some(targets) =>
+            //TODO: Add cause of renaming, requires FIRRTL change to RenameMap
+            targets.map { t => AnAnnotation(Some(t), Some(this), renames.associatedTransform.map(_.name)) }
+        }
+      } else Seq(this)
+    }
+
+    private def expand(stringBuilder: StringBuilder): StringBuilder = {
+      if (target.isDefined) {
+        stringBuilder.append(s"${target.get.serialize}")
+      } else {
+        stringBuilder.append(s"<DELETED>")
+      }
+      if (from.isDefined) {
+        val arrow = cause.map("(" + _ + ")").getOrElse("")
+        stringBuilder.append(s" <-$arrow- ")
+        from.get.expand(stringBuilder)
+      }
+      stringBuilder
+    }
+
+    override def serialize: String = expand(new StringBuilder()).toString
+  }
+
+  object StripDeleted extends Transform {
+
+    override def inputForm = UnknownForm
+
+    override def outputForm = UnknownForm
+
+    override def execute(a: CircuitState): CircuitState = {
+
+      val annotationsx = a.annotations.filter {
+        case a: DeletedAnnotation => false
+        case AnAnnotation(None, _, _) => false
+        case _: DupedResult => false
+        case _: DedupedResult => false
+        case _ => true
+      }
+
+      a.copy(annotations = annotationsx)
+
+    }
+
+  }
+
+  trait CircuitFixture {
+
+    /** An input FIRRTL string */
+    val input: String
+
+    lazy val output: String = input
+
+    /** Input annotations */
+    val annotations: AnnotationSeq = Seq.empty
+
+    val finalAnnotations: Option[AnnotationSeq] = None
+
+    lazy val state = CircuitState(Parser.parse(input), UnknownForm, annotations)
+  }
+
+  trait RightInverseFixture extends CircuitFixture {
+
+    /** An endomorphism */
```

Comment:
I think what this is getting at is that it is a function from the set of annotation sequences to the set of annotation sequences. Would it be possible to simplify some of the nomenclature to maybe "a function on an annotation sequence" to make it a bit more accessible?

Reply:
This languages dates from my original `MorphismSpec`. As far as I know the language is correct, but inaccessible. I'd probably prefer to keep it, but explain it. I was trying to be pedantic about the FP ideas here so that whatever the test was doing had some basis in a mathematical property.

Yeah, it's just saying that this is a mapping from `CircuitState => CircuitState`.